File: android/sample/src/main/kotlin/kmp/android/sample/vm/SampleViewModel.kt
package kmp.android.sample.vm

import kmp.android.shared.vm.BaseIntentViewModel
import kmp.android.shared.vm.VmEvent
import kmp.android.shared.vm.VmIntent
import kmp.android.shared.vm.VmState
import kmp.shared.analytics.domain.model.ToastAnalytics
import kmp.shared.analytics.domain.model.ToastAnalytics.ViewType
import kmp.shared.analytics.domain.usecase.TrackAnalyticsEventUseCase
import kmp.shared.analytics.domain.usecase.TrackAnalyticsEventUseCase.Params
import kmp.shared.base.ErrorResult
import kmp.shared.base.Result
import kmp.shared.sample.domain.model.SampleText
import kmp.shared.sample.domain.usecase.GetSampleTextUseCase

class SampleViewModel(
    private val getSampleText: GetSampleTextUseCase,
    private val trackAnalyticsEventUseCase: TrackAnalyticsEventUseCase,
) : BaseIntentViewModel<SampleState, SampleIntent, SampleEvent>(SampleState()) {

    override suspend fun applyIntent(intent: SampleIntent) {
        when (intent) {
            SampleIntent.OnAppeared -> loadSampleText()
            SampleIntent.OnButtonTapped -> showToast()
        }
    }

    private suspend fun loadSampleText() {
        update { copy(loading = true) }
        when (val result = getSampleText()) {
            is Result.Success -> update { copy(sampleText = result.data, loading = false) }
            is Result.Error -> update { copy(error = result.error, loading = false) }
        }
    }

    private suspend fun showToast() {
        trackAnalyticsEventUseCase(Params(ToastAnalytics.Event.presented(ViewType.Native)))

        _events.emit(SampleEvent.ShowMessage("Button was tapped"))
    }
}

data class SampleState(
    val loading: Boolean = false,
    val sampleText: SampleText? = null,
    val error: ErrorResult? = null,
) : VmState

sealed interface SampleIntent : VmIntent {
    data object OnAppeared : SampleIntent
    data object OnButtonTapped : SampleIntent
}

sealed interface SampleEvent : VmEvent {
    data class ShowMessage(val message: String) : SampleEvent
}

File: ios/PresentationLayer/Sample/Sources/Sample/Main/SampleViewModel.swift
//
//  Created by Petr Chmelar on 20.05.2022
//  Copyright © 2022 Matee. All rights reserved.
//

import DependencyInjection
import Factory
import KMPShared
import SharedDomain
import SharedDomainMocks
import SwiftUI
import UIToolkit

public final class SampleViewModel: UIToolkit.BaseViewModel, ViewModel, ObservableObject {
    
    @Injected(\.getSampleTextUseCase) private(set) var getSampleTextUseCase
    @Injected(\.trackAnalyticsEventUseCase) private(set) var trackAnalyticsEventUseCase
    
    // MARK: Lifecycle
    
    override public func onAppear() {
        super.onAppear()
        executeTask(Task { await loadSampleText() })
    }
    
    // MARK: State
    
    @Published public private(set) var state: State = State()

    public struct State {
        var sampleText: ViewData<SampleText> = .loading(mock: .stub)
        var toast: ToastData?
    }
    
    // MARK: Events
    
    public enum Event: ViewModelEvent {
        case showNextScreen
    }
    
    public var events: AsyncStream<Event> {
        AsyncStream { continuation in
            self._events = continuation
        }
    }
    
    private var _events: AsyncStream<Event>.Continuation?
    
    // MARK: Intent
    
    public enum Intent {
        case onButtonTapped
        case onToastChanged(data: ToastData?)
        case onNextTapped
    }

    public func onIntent(_ intent: Intent) {
        executeTask(Task {
            switch intent {
            case .onButtonTapped: showToast(message: "Button was tapped")
            case .onToastChanged(let data): state.toast = data
            case .onNextTapped: _events?.yield(.showNextScreen)
            }
        })
    }
    
    // MARK: Private
    
    private func loadSampleText() async {
        await execute {
            // Do the business logic
            state.sampleText = .data(try await getSampleTextUseCase.execute())
        } onError: { error in
            // Handle error
            state.sampleText = .error(error)
        } onCancel: { _ in
            // Custom cancel handling
        }
    }
    
    private func showToast(message: String) {
        Task {
            try? await trackAnalyticsEventUseCase.invoke(
                params: TrackAnalyticsEventUseCaseParams(event: ToastAnalytics.ToastPresentedEvent(viewType: .native))
            )
        }
        
        state.toast = ToastData(message, hideAfter: 2)
    }
}
